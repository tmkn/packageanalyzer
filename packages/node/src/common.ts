import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
