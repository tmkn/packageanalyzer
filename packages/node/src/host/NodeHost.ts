import { Writable } from "node:stream";

import type { IWriter, IHost } from "../../../shared/src/host/IHost.js";
import { npmOnline } from "../providers/online.js";
import { OraLogger } from "../loggers/OraLogger.js";
import type { IPackageJsonProvider } from "../../../shared/src/providers/provider.js";
import type { ILogger } from "../../../shared/src/loggers/ILogger.js";

export class NodeWriter implements IWriter {
    private readonly writer: WritableStreamDefaultWriter<string>;
    private pendingWrites: Promise<void>[] = [];

    constructor(nodeStream: Writable) {
        const webStream = Writable.toWeb(nodeStream);
        this.writer = webStream.getWriter();
    }

    write(chunk: string): Promise<void> {
        const writePromise = this.writer.write(chunk);

        this.pendingWrites.push(writePromise);

        return writePromise;
    }

    async flush(): Promise<void> {
        await Promise.all(this.pendingWrites);

        this.pendingWrites = [];
    }
}

export class NodeHost implements IHost {
    private readonly stdoutWriter: IWriter;
    private readonly stderrWriter: IWriter;

    constructor(
        private readonly _stdout: Writable,
        private readonly _stderr: Writable
    ) {
        this.stdoutWriter = new NodeWriter(this._stdout);
        this.stderrWriter = new NodeWriter(this._stderr);
    }

    getStdoutWriter(): IWriter {
        return this.stdoutWriter;
    }

    getStderrWriter(): IWriter {
        return this.stderrWriter;
    }

    getDefaultProvider(): IPackageJsonProvider {
        return npmOnline;
    }

    getLogger(): ILogger {
        return new OraLogger();
    }
}
