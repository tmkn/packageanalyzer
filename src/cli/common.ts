import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
