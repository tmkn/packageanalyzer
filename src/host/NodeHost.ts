import { Writable } from "node:stream";

import type { IWriter, IHost } from "./IHost.js";
import type { ILogger, IPackageJsonProvider } from "../index.web.js";
import { npmOnline } from "../providers/online.js";
import { OraLogger } from "../loggers/OraLogger.js";

export class NodeWriter implements IWriter {
    private readonly writer: WritableStreamDefaultWriter<string>;
    private pendingWrites: Promise<void>[] = [];

    constructor(nodeStream: Writable) {
        const webStream = Writable.toWeb(nodeStream);
        this.writer = webStream.getWriter();
    }

    write(chunk: string): Promise<void> {
        const writePromise = this.writer.write(chunk);

        this.pendingWrites.push(writePromise);

        return writePromise;
    }

    async flush(): Promise<void> {
        await Promise.all(this.pendingWrites);

        this.pendingWrites = [];
    }
}

export class NodeHost implements IHost {
    private readonly stdoutWriter: IWriter;
    private readonly stderrWriter: IWriter;

    constructor(
        private readonly _stdout: Writable,
        private readonly _stderr: Writable
    ) {
        this.stdoutWriter = new NodeWriter(this._stdout);
        this.stderrWriter = new NodeWriter(this._stderr);
    }

    getStdoutWriter(): IWriter {
        return this.stdoutWriter;
    }

    getStderrWriter(): IWriter {
        return this.stderrWriter;
    }

    getDefaultProvider(): IPackageJsonProvider {
        return npmOnline;
    }

    getLogger(): ILogger {
        return new OraLogger();
    }
}
