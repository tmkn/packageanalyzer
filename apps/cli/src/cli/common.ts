import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
