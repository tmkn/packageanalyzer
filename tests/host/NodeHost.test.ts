import { describe, test, expect } from "vitest";

import { NodeWriter, NodeHost } from "../../src/host/NodeHost.js";
import { npmOnline, OraLogger } from "../../src/index.js";
import { createMockContext } from "../common.js";

describe(`NodeAsyncWriter Tests`, () => {
    test(`writes out all data`, async () => {
        const { stdout } = createMockContext();
        const nodeWriter = new NodeWriter(stdout);

        nodeWriter.write(`line1\n`);
        nodeWriter.write(`line2\n`);

        await nodeWriter.flush();

        expect(stdout.lines).toMatchSnapshot();
    });
});

describe(`NodeHost Tests`, () => {
    test(`returns the same stdout writer`, () => {
        const { stdout, stderr } = createMockContext();
        const host = new NodeHost(stdout, stderr);

        const writer1 = host.getStdoutWriter();
        const writer2 = host.getStdoutWriter();

        expect(writer1).toBe(writer2);
    });

    test(`returns the same stderr writer`, () => {
        const { stdout, stderr } = createMockContext();
        const host = new NodeHost(stdout, stderr);

        const writer1 = host.getStderrWriter();
        const writer2 = host.getStderrWriter();

        expect(writer1).toBe(writer2);
    });

    test(`returns npmOnline as default provider`, () => {
        const { stdout, stderr } = createMockContext();
        const host = new NodeHost(stdout, stderr);

        const defaultProvider = host.getDefaultProvider();

        expect(defaultProvider).toBe(npmOnline);
    });

    test(`returns OraLogger as logger`, () => {
        const { stdout, stderr } = createMockContext();
        const host = new NodeHost(stdout, stderr);

        const logger = host.getLogger();

        expect(logger).toBeInstanceOf(OraLogger);
    });
});
