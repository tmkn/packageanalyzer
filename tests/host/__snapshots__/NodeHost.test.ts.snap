// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NodeAsyncWriter Tests > writes out all data 1`] = `
[
  "line1",
  "line2",
  "",
]
`;
